include/deal.II/grid/grid_tools.h

```diff
@@ -1285,6 +1389,272 @@ namespace GridTools
+       * Pointers of ranges to found intersections for requested intersection.
+       */
+      std::vector<unsigned int> recv_ptrs;
+
+      /**
+       * Ranks from where data is received.
+       */
+      std::vector<unsigned int> recv_ranks;
+
+      /**
+       * Distribute quadrature points to found intersections and construct
+       * GridTools::internal::DistributedComputePointLocationsInternal from
+       * class members. This can be done without searching for points agains
+       * since all information is locally known.
+       */
+      GridTools::internal::DistributedComputePointLocationsInternal<dim,
+                                                                    spacedim>
+      convert_to_distributed_compute_point_locations_internal(
+        const unsigned int                  n_quadrature_pnts,
+        const Triangulation<dim, spacedim> &tria,
+        const Mapping<dim, spacedim> &      mapping) const
+      {
+        using CellIterator =
+          typename Triangulation<dim, spacedim>::active_cell_iterator;
+
+        GridTools::internal::DistributedComputePointLocationsInternal<dim,
+                                                                      spacedim>
+          result;
+
+
+        // We need quadrature rules for the intersections. We are using a
+        // QGaussSimplex quadrature rule since CGAL always returns simplices as
+        // intersections.
+        const QGaussSimplex<dim - 1> quadrature(n_quadrature_pnts);
+
+        // Resulting quadrature points get different indices. In the case the
+        // requested intersections are unique also the resulting quadrature
+        // points are unique and we can simply number the points in an
+        // ascending way.
+        for (auto const &recv_component : recv_components)
+          {
+            // dependent on the size of the intersection an empty quadrature is
+            // returned. Therefore, we have to compute the quadrature also here.
+            const auto &quad = quadrature.compute_affine_transformation(
+              std::get<2>(recv_component));
+
+            for (unsigned int i = 0; i < quad.size(); ++i)
+              {
+                result.recv_components.emplace_back(
+                  std::get<0>(recv_component),
+                  result.recv_components.size(), // number of point
+                  numbers::invalid_unsigned_int);
+              }
+          }
+
+        // since empty quadratures might be present we have to set the number of
+        // searched points after inserting the point indices into
+        // recv_components
+        result.n_searched_points = result.recv_components.size();
+
+#ifdef DEAL_II_WITH_MPI
+        // indices assigned at recv side needed to fill send_components
+        const std::map<unsigned int, std::vector<unsigned int>>
+          indices_of_rank = communicate_indices(result.recv_components,
+                                                tria.get_communicator());
```

Comment:
I don't yet get why the communication is needed here. What information is available at this stage? The sending side is the slave side and the receiving side is the master side, aren't they?

Reply:
You are right. The sending side reports to the receiving side. Therefore, all information is present at the receiving side. Especially:
* Which intersection has been found on which rank

The sending side only knows about intersections found on the corresponding rank and does not know if an intersection is also found on another sending rank. The numbering in ```DistributedComputePointLocationsInternal``` on the receiver side is therefore clear. On the sending side, however, it is not clear which number each entry has. E.g. the indices could start at 60 if 60 other quadrature points are found on a rank with smaller index. In general, we can not just work with an offset because intersection 1,3,5 might be found on the sending rank under consideration and 2,4 might be found on another rank (or multiple intersections for the same request might be found on different ranks). Therefore, I send away all the indices determined on the receiver side to the senders and assign these indices.

Reply:
See discussion below https://github.com/dealii/dealii/pull/15165#discussion_r1186853645